From 16c3ea785fe4a383c6675dfe7316f3c815755bdd Mon Sep 17 00:00:00 2001
From: Brandon Philips <bphilips@suse.de>
Date: Wed, 15 Sep 2010 09:24:24 +0000
Subject: [PATCH] net: enable GRO by default for vlan devices
References: bnc#582730
Patch-mainline: 2.6.37? Merged into net-next-2.6

Currently vlan devices don't have GRO by default as none of the Ethernet
drivers add NETIF_F_GRO to their vlan_features.

As GRO is a software feature add GRO to dev->vlan_features in
register_netdevice() and let vlan_dev_init() take care that it gets
enabled only when dev->features has NETIF_F_GRO too.

Signed-off-by: Brandon Philips <bphilips@suse.de>
Acked-by: Eric Dumazet <eric.dumazet@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 net/core/dev.c |    5 +++++
 1 file changed, 5 insertions(+)

Index: linux-2.6.32-SLE11-SP1/net/core/dev.c
===================================================================
--- linux-2.6.32-SLE11-SP1.orig/net/core/dev.c
+++ linux-2.6.32-SLE11-SP1/net/core/dev.c
@@ -4884,6 +4884,11 @@ int register_netdevice(struct net_device
 	if (dev->features & NETIF_F_SG)
 		dev->features |= NETIF_F_GSO;
 
+	/* Enable GRO for vlans by default if dev->features has GRO also.
+	 * vlan_dev_init() will do the dev->features check.
+	 */
+	dev->vlan_features |= NETIF_F_GRO;
+
 	netdev_initialize_kobject(dev);
 	ret = netdev_register_kobject(dev);
 	if (ret)
